File: src/2179.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#ifndef FENWICKTREE_H
#define FENWICKTREE_H

typedef struct {
    int* tree;
    int size;
} FenwickTree;
FenwickTree* fenwickTreeCreate(int size) {
    FenwickTree* obj = NULL;

    obj = (FenwickTree*)malloc(sizeof(FenwickTree));
    if (obj == NULL) {
        perror("malloc");
        return obj;
    }

    obj->size = size;
    obj->tree = (int*)calloc(size + 1, sizeof(int));
    if (obj->tree == NULL) {
        perror("malloc");
        free(obj);
        obj = NULL;
        return obj;
    }

    return obj;
}
void fenwickTreeUpdate(FenwickTree* obj, int index, int delta) {
    index++;
    while (index <= obj->size) {
        obj->tree[index] += delta;
        index += (index & -index);
    }
}
int fenwickTreeQuery(FenwickTree* obj, int index) {
    int retVal = 0;

    index++;
    while (index > 0) {
        retVal += obj->tree[index];
        index -= (index & -index);
    }

    return retVal;
}
void fenwickTreeFree(FenwickTree* obj) {
    free(obj->tree);
    obj->tree = NULL;
    free(obj);
    obj = NULL;
}

#endif  // FENWICKTREE_H
long long goodTriplets(int* nums1, int nums1Size, int* nums2, int nums2Size) {
    long long retVal = 0;

    int* pos2 = (int*)malloc(nums1Size * sizeof(int));
    if (pos2 == NULL) {
        perror("malloc");
        return retVal;
    }
    for (int i = 0; i < nums1Size; i++) {
        pos2[nums2[i]] = i;
    }

    int* reversedIndexMapping = (int*)malloc(nums1Size * sizeof(int));
    if (reversedIndexMapping == NULL) {
        perror("malloc");
        free(pos2);
        return retVal;
    }
    for (int i = 0; i < nums1Size; i++) {
        reversedIndexMapping[pos2[nums1[i]]] = i;
    }

    FenwickTree* tree = fenwickTreeCreate(nums1Size);
    if (tree == NULL) {
        perror("malloc");
        free(pos2);
        free(reversedIndexMapping);
        return retVal;
    }

    for (int value = 0; value < nums1Size; value++) {
        int pos = reversedIndexMapping[value];
        int left = fenwickTreeQuery(tree, pos);
        fenwickTreeUpdate(tree, pos, 1);

        int right = (nums1Size - 1 - pos) - (value - left);
        retVal += ((long long)left * right);
    }

    // Free allocated memory
    free(pos2);
    free(reversedIndexMapping);
    fenwickTreeFree(tree);

    return retVal;
}

int main(int argc, char** argv) {
#define MAX_SIZE (int)(1e5)
    struct testCaseType {
        int nums1[MAX_SIZE];
        int nums1Size;
        int nums2[MAX_SIZE];
        int nums2Size;
    } testCase[] = {{{2, 0, 1, 3}, 4, {0, 1, 2, 3}, 4}, {{4, 0, 1, 3, 2}, 5, {4, 1, 0, 2, 3}, 5}};
    int numberOfTestCase = sizeof(testCase) / sizeof(testCase[0]);
    /* Example
     *  Input: nums1 = [2,0,1,3], nums2 = [0,1,2,3]
     *  Output: 1
     *
     *  Input: nums1 = [4,0,1,3,2], nums2 = [4,1,0,2,3]
     *  Output: 4
     */

    long long answer;
    int i, j;
    for (i = 0; i < numberOfTestCase; ++i) {
        printf("Input: nums1 = [");
        for (j = 0; j < testCase[i].nums1Size; ++j) {
            printf("%s%d", (j == 0) ? "" : ",", testCase[i].nums1[j]);
        }
        printf("], nums2 = [");
        for (j = 0; j < testCase[i].nums2Size; ++j) {
            printf("%s%d", (j == 0) ? "" : ",", testCase[i].nums2[j]);
        }
        printf("]\n");

        answer = goodTriplets(testCase[i].nums1, testCase[i].nums1Size, testCase[i].nums2, testCase[i].nums2Size);
        printf("Output: %lld\n", answer);

        printf("\n");
    }

    return EXIT_SUCCESS;
}

File: src/2179.py
import logging
import sys
from bisect import bisect_left, bisect_right
from collections import Counter, defaultdict, deque
from functools import cache, lru_cache
from heapq import heapify, heappop, heappush
from itertools import accumulate
from typing import List, Optional


def logging_setting():
    LOG_LEVEL = logging.INFO    # DEBUG, INFO, WARNING, ERROR, CRITICAL
    if __debug__:
        LOG_LEVEL = logging.DEBUG

    logging.basicConfig(
        stream=sys.stderr,
        level=LOG_LEVEL,
        format="%(levelname)-6s %(asctime)s %(module)s:%(funcName)s:%(lineno)-4s %(message)s",
        datefmt='%Y/%m/%d %H:%M:%S')


class FenwickTree:
    def __init__(self, size: int) -> None:
        self.tree = [0] * (size + 1)

    def update(self, index: int, delta: int) -> None:
        index += 1
        treeSize = len(self.tree)
        while index <= treeSize - 1:
            self.tree[index] += delta
            index += (index & -index)

    def query(self, index: int) -> int:
        retVal = 0

        index += 1
        while index > 0:
            retVal += self.tree[index]
            index -= (index & -index)

        return retVal


class Solution:
    def goodTriplets(self, nums1: List[int], nums2: List[int]) -> int:
        retVal = 0

        nums1Size = len(nums1)

        pos2 = [0] * nums1Size
        for i, num2 in enumerate(nums2):
            pos2[num2] = i

        reversedIndexMapping = [0] * nums1Size
        for i, num1 in enumerate(nums1):
            reversedIndexMapping[pos2[num1]] = i

        tree = FenwickTree(nums1Size)

        for value in range(nums1Size):
            pos = reversedIndexMapping[value]
            left = tree.query(pos)
            tree.update(pos, 1)
            right = (nums1Size - 1 - pos) - (value - left)
            retVal += (left * right)

        return retVal


if __name__ == "__main__":
    logging_setting()

    try:
        logging.info("sys.version: %s", sys.version)
        print()

        pSolution = Solution()
        for nums1, nums2 in zip([[2, 0, 1, 3], [4, 0, 1, 3, 2]], [[0, 1, 2, 3], [4, 1, 0, 2, 3]]):
            # /* Example
            #  *  Input: nums1 = [2,0,1,3], nums2 = [0,1,2,3]
            #  *  Output: 1
            #  *
            #  *  Input: nums1 = [4,0,1,3,2], nums2 = [4,1,0,2,3]
            #  *  Output: 4
            #  */
            logging.info("Input: nums1 = %s, nums2 = %s", nums1, nums2)

            retVal = pSolution.goodTriplets(nums1, nums2)
            logging.info("Output: %s", retVal)

            print()
    except KeyboardInterrupt as exception:
        logging.error("%s: %s", exception.__class__.__name__, exception, exc_info=True)
        pass
    except Exception as exception:
        logging.error("%s: %s", exception.__class__.__name__, exception, exc_info=True)
        pass

    sys.exit(0)
